### PR TITLE
[docs] DeepSpeed

### PR DESCRIPTION
**Pull request by stevhliu**
Refactors the [DeepSpeed API page](https://huggingface.co/docs/transformers/main/en/main_classes/deepspeed#deepspeed-integration) to make it easier to find and view `HfDeepSpeedConfig`, the only actual API reference on this doc. The rest of the DeepSpeed content will go to the Efficient Training Techniques section as a standalone guide. I've also moved some of the troubleshooting content with building DeepSpeed to the [Debugging guide](https://huggingface.co/docs/transformers/main/en/debugging) with a link.

todo:

- [x] discuss choosing which ZeRO stage to use
- [x] get model weights out
- [x] ZeRO-3 and inference
- [x] memory requirements
- [x] troubleshooting/filing issues
- [x] non-Trainer DeepSpeed integration

### CODE REVIEW

**Comment by stevhliu** [comment]
First draft wrapped up!

1. It is easier to find `HfDeepSpeedConfig` now
2. Prebuilding troubleshooting steps are in `debugging.md`
3. Testing the DeepSpeed integration is moved to `testing.md`
4. The main DeepSpeed guide is refactored to flow better. For example, installation > memory requirements > selecting a ZeRO stage. I think it's important to address these first before a user finds out they don't have the required memory or chose the wrong ZeRO stage. I've also condensed and made things more concise where appropriate without losing any context or info.

**Review by amyeroberts — APPROVED** [review]
Look great - thanks for all the work making the docs easier to navigate 🧭 

cc @pacman100 to confirm the details are correct

File: docs/source/en/testing.md [hunk 1]
```diff
@@ -1312,3 +1312,19 @@ You can vote for this feature and see where it is at these CI-specific threads:
...
 - [CircleCI:](https://ideas.circleci.com/ideas/CCI-I-344)
+
+## DeepSpeed integration
+
+For a PR that involves the DeepSpeed integration, keep in mind our CircleCI PR CI setup doesn't have GPUs. Tests requiring GPUs are run on a different CI nightly. This means if you get a passing CI report in your PR, it doesn’t mean the DeepSpeed tests pass.
+
+To run DeepSpeed tests:
+
+```bash
+RUN_SLOW=1 pytest tests/deepspeed/test_deepspeed.py
```
Comment (pacman100) [inline docs/source/en/testing.md:1323]: Nice for providing these details!

**Review by pacman100 — APPROVED** [review]
Thank you @stevhliu for making it easier to get documentation for `HfDeepSpeedConfig` class in the main classes section. Moving the usage details to the section on `Efficient Training Techniques` makes it inline with other similar method docs such as FSDP, DDP ✨.

**Pull request merged**